src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java

```diff
@@ -50,26 +102,122 @@ public FluidStack getFluid() {
         return null;
     }
 
+    public Object[] getProperty(String PropertyName) {
+        return getProperty(PropMap.get(PropertyName));
+    }
+
+    public Object[] getProperty(stockProperty PropertyName) {
+        if (stock != null) {
+            switch(PropertyName) {
+                case ID:
+                    return new Object[] { stock.getDefinitionID() };
+                case NAME:
+                    return new Object[] { stock.getName() };
+                case TAG:
+                    return new Object[] { stock.tag };
+                case WEIGET:
+                    return new Object[] { stock.getWeight() };
+                case SPEED:
+                    if (stock instanceof EntityMoveableRollingStock) {
+                        return new Object[] { (EntityMoveableRollingStock)stock.getCurrentSpeed().metric() };
+                    } 
+                    return null;
+                case DIRECTION:
+                    if (stock instanceof EntityMoveableRollingStock) {
+                        double[] SpeedVector = getSpeedVector();
+                        return new Object[] { EnumFacing.fromAngle(SpeedVector[1]).toString()) };
+                    } else {
+                        EnumFacing dir = EnumFacing.fromAngle(stock.getRotationYaw());
```

Comment:
I don't think you need the if/else here.  The else condition seems to be correct as the speedVector is generated from the rotation yaw value

Reply:
I think it is necessary.
In your info() function, when the currentSpeed < 0, your code set the direction opposite through:
 dir = dir.getOpposite();

My function getSpeedVector, do the same thing in the if condition; and fall back to else condition when it don't belong to EntityMoveableRollingStock.

Reply:
Looking at how the code is written, EnumFacing.fromAngle(stock.getRotationYaw()) is the correct value.

Reply:
The other code is wrong

Reply:
> The other code is wrong
![image](https://user-images.githubusercontent.com/7941967/66263514-8e945d80-e826-11e9-99e2-f653fb543231.png)
Cam, this is your origin code. 
https://github.com/TeamOpenIndustry/ImmersiveRailroadingIntegration/blob/forge_1.12.2/src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java#L63
if the metric()<0 , dir = dir.getOpposite();
Maybe it is my misunderstanding, about the getOpposite. What should I do? Delete the if condition and leave the else part alone?